### ISSUE TITLE
New Function / Merge -- Network Connectedness

### ISSUE THREAD

**Issue by ZhijieWang** [networkx/networkx#1953]
Hi,

I used the Networkx library to help a marketing professor to analyze social network data. She was interested in the connectedness property of the graph. She listed a R package, however other project related code are written in python. So I hacked my way with a quick implementation using all_pairs_node_connectivity. If it is valuable to networkx project, I can cleanup and make a pull request.

Quote:
– Krackhardt’s connectedness for a digraph (directed graph) G is
equal to the fraction of all dyads (a group of two nodes), u and v,
such that there exists an undirected path from u to v in G

R package in discussion is from here, specifically page 11:
http://www.csc.ncsu.edu/faculty/samatova/practical-graph-mining-with-R/slides/pdf/Link_Analysis.pdf

Current non-robust implementation:

        all_pairs_connectedness = nx.all_pairs_node_connectivity(G)

        ## count all pairs with connectivity value > 0
        count = 0

        for u in all_pairs_connectedness.values():
                count+=(sum(x> 0 for x in u.values()))
        ## Use 1.0 instead of 1 to force value conversion from int to float.
        ## Otherwise integer division will not have decimal portion
        connectedness = count / (G.number_of_nodes()*(G.number_of_nodes()-1.0))

**Comment by jfinkels** [comment]
I think your implementation is doing more work than is necessary, since "connectivity" is computing the number of nodes required to disconnect the graph, which is more information than you need; you only need reachability of pairs of nodes.

There are several options for implementing this. One option is to compute the density of the reachability graph, as described in the documentation for that R function. (`reachability_graph` could become its own function, there are various interesting implementations for it.)

    # Could alternately not compute H here,
    # and just call `G.has_path(u, v) or G.has_path(u, v)` below.
    H = nx.Graph(G)
    reachability_graph = nx.Graph((u, v) for u, v in combinations(H, 2) if H.has_path(u, v))
    connectedness = nx.density(reachability_graph)

Another option is to compute the numerator and denominator directly.

    num_connected_diads = sum(1 for u, v in combinations(G, 2) if G.has_path(u, v) or G.has_path(v, u))
    n = len(G)
    num_diads = (n * (n - 1)) // 2
    connectedness = num_connected_diads / num_diads

**Comment by ZhijieWang** [comment]
Thank you @jfinkels  for point me to the right direction.  I am pretty new to the networkx library.

**Comment by dschult** [comment]
Can you find the connected components (and thus their size N_k) and then use len(G) and N_k?
Probably cheaper than recomputing has_path many times.

**Comment by jfinkels** [comment]
Yes you could also do that, but you'll have to make sure to get the arithmetic right for counting the number of undirected pairs that are not connected :)

**Comment by dschult** [comment]
Yes, it's true.... getting the arithmetic right is necessary for counting undirected pairs.    :)

**Comment by ZhijieWang** [comment]
Thanks all the input. Analyzing a grpah of 5M+ nodes
1. Using my original implementation, OOM KILLED
2. Using direct calculation of numerator and denominator --32 hours still on going.

I guess i have to go parallel.

**Comment by jfinkels** [comment]
Can you provide your new implementation? It still may be valuable.

**Comment by ZhijieWang** [comment]
Not a python power user or concurrent python programmer, but this is what I have for now.
Intended to calculate large graph, so I used multithreading, for shared memory. Using ``concurrency.future`` for GIL work-around. Currently running performance testing on My mac mini 2011 ( i5 CPU 16 GB RAM). Will update with result soon.
```
import networkx as nx
import concurrent.futures
import itertools
import time
from functools import reduce
def base_case():
    # Original test case data taken from #http://www.csc.ncsu.edu/faculty/samatova/practical-graph-mining-with-R/slides/pdf/Link_Analysis.pdf
    # Connectedness should be 0.8
    G = nx.Graph()
    G.add_edge(1,2)
    G.add_edge(1,3)
    G.add_edge(1,4)
    G.add_edge(3,9)
    G.add_edge(9,8)
    G.add_edge(9,6)
    G.add_edge(6,4)
    G.add_edge(4,5)
    G.add_edge(4,5)
    G.add_edge(7,5)
    G.add_node(10)
    return G
def conncected(n):
    return nx.has_path(G,n[0],n[1])
def concurrent_connectedness():
    with concurrent.futures.ThreadPoolExecutor(max_workers=500) as executor:
        num_connected_diads =sum(executor.map(
        conncected,itertools.combinations(G,2)
            )
        )
        n =len(G)
        num_diads = (n * (n - 1)) // 2
        connectedness = num_connected_diads / num_diads
        return connectedness
def test_performance(n):
    global G
    G = nx.fast_gnp_random_graph(n,.6)
    concurrent_connectedness()
def singlethread_connectedness(G):
    num_connected_diads = sum(1 for u, v in itertools.combinations(G, 2) if nx.has_path(G,u, v) or nx.has_path(G,v, u))
    n = len(G)
    num_diads = (n * (n - 1)) // 2
    connectedness = num_connected_diads / num_diads
    return connectedness
def main():
    global G
    for n in [10,100,1000]:
        elapsed=list()
        for i in range(100):
            start=time.time()
            test_performance(n)
            elapsed.append(time.time()-start)
        print("Average run time for size of "+str(n)+" is :" )
        print(reduce(lambda x, y: x + y, elapsed) / len(elapsed))



if __name__=='__main__':
    main()

```

**Comment by jfinkels** [comment]
You still have not tried the implementation that @dschult suggested, which is likely to be much faster: compute the size of each (weakly) connected component in the graph, then compute the total number of pairs of nodes within the same component and the total number of pairs of nodes in different components.

**Comment by ZhijieWang** [comment]
@jfinkels That't true. weakly connected component is a BFS rather than exhaustedly verify pairs (probably by bfs again).

**Comment by ZhijieWang** [comment]
Uhh, tried with ``reachability_graph`` approach. Unfortunately, it doesn't seem to be able to scale to 5 graph of M+ nodes.

**Comment by dschult** [comment]
Try ```del G``` after you are done computing ```nx.connected_components```.
That will clear out memory for your next graph.

**Comment by jfinkels** [comment]
In any case, it seems this function can be implemented as

    def connectedness(G):
        n = len(G)
        lengths = map(len, nx.weakly_connected_components(G))
        # We are double counting both sets of dyads, but that's okay since both the numerator
        # and the denominator have a factor of two.
        num_connected_dyads = sum(m * (m - 1) for m in lengths)
        num_dyads = n * (n - 1)
        return num_connected_dyads / num_dyads

Someone double-check my math?

**Comment by ZhijieWang** [comment]
@jfinkels Your math is correct, I tested out. Your counting is similar to mine. 
My graph is undirected, so I used connected_components(G)

@dschult , I doubt `del G` will be of much help. Since I run into CPU bound, I booted up my workstation to run it. 32GB DDR3 with i7-4790k. The graph was original generated from a two-way table (python pandas.DataFrame). The dataframe was deleted immediately after original graph construction. And I did not use a subgraph then density approach to calculate connectedness. I have a ``dyads`` counting using ``len``. Memory stress during the entire process is less than 40%. ( I suspect the graph is highly connected after sampling several subgraph)

Since the original data is a two-way table of x by y, with x> 5M+ and y<50. I can coarsen the large graph G into a subgraph G' of only nodes from y. If there is a Xi such that Xi is connected with U and V, given U and V are from y, then everything connected to U are connected to V. A reachability_components on  G' will serve the purpose of BFS the G. However, it is going to be tricky to count dyads ( merging all nodes of each sub connected components into a set? need to test it out)
